File: backend/app/api/routes/order.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

import schemas
import services
from app.db import get_db
from app.core.permissions import require_permission

router = APIRouter(prefix="/orders", tags=["orders"])

@router.post("")
async def create_order(
    order: schemas.OrderCreate,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:create"))
):
    return await services.create_order(db, order, current_user)

@router.get("/{order_id}")
async def get_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:read"))
):
    return await services.get_order(db, order_id, current_user)

@router.get("")
async def list_orders(
    user_id: int | None = None,
    status: str | None = None,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:list"))
):
    return await services.list_orders(db, user_id, status, current_user)

@router.post("/{order_id}/confirm")
async def confirm_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:update"))
):
    return await services.confirm_order(db, order_id, current_user)


@router.post("/{order_id}/check-in")
async def check_in_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:update"))
):
    return await services.check_in_order(db, order_id, current_user)


@router.post("/{order_id}/check-out")
async def check_out_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:update"))
):
    return await services.check_out_order(db, order_id, current_user)

@router.post("/{order_id}/cancel")
async def cancel_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:update"))
):
    return await services.cancel_order(db, order_id, current_user)


@router.post("/{order_id}/refund")
async def refund_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:refund"))
):
    return await services.refund_order(db, order_id, current_user)


@router.post("/{order_id}/extend")
async def extend_order(
    order_id: int,
    payload: schemas.OrderExtend,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:update"))
):
    return await services.extend_order(db, order_id, payload.extra_days, current_user)


@router.post("/{order_id}/recalculate")
async def recalculate_order(
    order_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_permission("order:update"))
):
    return await services.recalculate_order(db, order_id, current_user)

File: frontend/src/api/order.ts
import http from "./http";
import type { OrderResponse, OrderStatus } from "@/types";

export interface OrderCreatePayload {
  name: string;
  phone_number: string;
  room_number: string;
  stay_length: number;
}

export interface OrderExtendPayload {
  extra_days: number;
}

export interface OrderListParams {
  user_id?: number;
  status?: OrderStatus | "";
}

function buildListParams(params: OrderListParams = {}) {
  const query: Record<string, number | string> = {};

  if (params.user_id !== undefined) {
    query.user_id = params.user_id;
  }

  if (params.status) {
    query.status = params.status;
  }

  return query;
}

export async function createOrder(payload: OrderCreatePayload): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>("/orders", payload);
  return res.data;
}

export async function getOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.get<OrderResponse>(`/orders/${orderId}`);
  return res.data;
}

export async function listOrders(params: OrderListParams = {}): Promise<OrderResponse[]> {
  const res = await http.get<OrderResponse[]>("/orders", {
    params: buildListParams(params),
  });
  return res.data;
}

export async function confirmOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/confirm`);
  return res.data;
}

export async function checkInOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/check-in`);
  return res.data;
}

export async function checkOutOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/check-out`);
  return res.data;
}

export async function cancelOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/cancel`);
  return res.data;
}

export async function refundOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/refund`);
  return res.data;
}

export async function extendOrder(
  orderId: number,
  payload: OrderExtendPayload,
): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/extend`, payload);
  return res.data;
}

export async function recalculateOrder(orderId: number): Promise<OrderResponse> {
  const res = await http.post<OrderResponse>(`/orders/${orderId}/recalculate`);
  return res.data;
}
